internal/core/src/storage/FileWriter.h

```diff
@@ -0,0 +1,134 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cassert>
+#include <stdexcept>
+#include <string>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <cstring>
+#include <thread>
+
+#include "common/EasyAssert.h"
+#include "storage/ThreadPools.h"
+
+namespace milvus::storage {
+
+class FileWriter {
+ public:
+    explicit FileWriter(std::string filename);
+
+    ~FileWriter();
+
+    void
+    Write(const void* data, size_t size);
+
+    size_t
+    Finish();
+
+ private:
+    static constexpr size_t ALIGNMENT_BYTES = 4096;
+    static constexpr size_t ALIGNMENT_MASK = ALIGNMENT_BYTES - 1;
+
+    void
+    WriteWithDirectIO(const void* data, size_t nbyte);
+
+    void
+    WriteWithBufferedIO(const void* data, size_t nbyte);
+
+    bool
+    PositionedWrite(const void* data, size_t nbyte, size_t file_offset);
+
+    void
+    Cleanup() noexcept;
+
+    int fd_{-1};
+    std::string filename_{""};
+    size_t file_size_{0};
+
+    // for direct io
+    bool use_direct_io_{false};
+    void* aligned_buf_{nullptr};
+    size_t capacity_{0};
+    size_t offset_{0};
+};
+
+class FileWriterConfig {
```

Comment:
It pretty weird to me to have a singleton here. Why we just use a global parameters read from yml file?